### PR TITLE
BUG: clip doesn't preserve dtype by column

### LINKED ISSUE #24162: DataFrame.clip_upper does not preserve dtype per column
#### Code Sample

```python
import pandas as pd
data = pd.DataFrame({'INT': [-1, 0, 10, 9],
              'FLOAT': [-0.148, 0.2347, 38.237, 12.2233]},
             index=pd.date_range("20180101 00:00", periods=4))

print('Original data:')
print(data.head())

print('\nThis is probably not a bug but my misunderstanding:')
print('(So how would I apply "clip_upper" inplace on parts of the dataframe?)')
data.loc[[True, True, True, False], ['INT']].clip_upper(8, inplace=True)
print(data.head()) 
# I used then:
# data.loc[[True, True, True, False], ['INT']] = data.loc[[True, True, True, False], ['INT']].clip_upper(8)     

print('\nIt seems that clip_upper does not preserve the dtypes:')
print(data.clip_upper(8).head())

print('\nSame for inplace:')
data.clip_upper(8, inplace=True)
print(data.head())
```
##### Output of this code:

<details>

```
Original data:
            INT    FLOAT
2018-01-01   -1  -0.1480
2018-01-02    0   0.2347
2018-01-03   10  38.2370
2018-01-04    9  12.2233

(A) This is probably not a bug but my misunderstanding:
(So how would I apply "clip_upper" inplace on parts of the dataframe?)
            INT    FLOAT
2018-01-01   -1  -0.1480
2018-01-02    0   0.2347
2018-01-03   10  38.2370
2018-01-04    9  12.2233

(B) It seems that clip_upper does not preserve the dtypes:
            INT   FLOAT
2018-01-01 -1.0 -0.1480
2018-01-02  0.0  0.2347
2018-01-03  8.0  8.0000
2018-01-04  8.0  8.0000

(C) Same for inplace:
            INT   FLOAT
2018-01-01 -1.0 -0.1480
2018-01-02  0.0  0.2347
2018-01-03  8.0  8.0000
2018-01-04  8.0  8.0000
```

</details>

#### Problem description

`clip_upper` with int- and float- columns convert int-column to float.

Calling `data.clip_upper(10)` with an integer, I would expect that it leaves the int-column as integers and the float-column as float. However, it converts everything to float. (see (B) and (C))

Moreover, `clip_upper` with `inplace=True` does not work with `.loc` but this might as well be me understanding the concept wrong... (see (A))

Same for `clip_lower`.

#### Expected Output
For (A):
```
            INT    FLOAT
2018-01-01   -1  -0.1480
2018-01-02    0   0.2347
2018-01-03    8  38.2370
2018-01-04    9  12.2233
```

For (B) and (C):
```
            INT   FLOAT
2018-01-01 -1 -0.1480
2018-01-02  0  0.2347
2018-01-03  8  8.0000
2018-01-04  8  8.0000
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None

pandas: 0.23.4
pytest: 4.0.1
pip: 18.1
setuptools: 40.6.2
Cython: 0.29
numpy: 1.15.4
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 7.2.0
sphinx: 1.8.2
patsy: None
dateutil: 2.7.5
pytz: 2018.7
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.1
openpyxl: 2.5.11
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.1.2
lxml: 4.2.5
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: 1.2.14
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by minggli**
- [x] closes #24162
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

```python3
>>> import pandas as pd
>>> data = pd.DataFrame([[1, 2], [3, 4]], columns=['int1', 'int2'])
>>> data.clip_upper(1)
__main__:1: FutureWarning: clip_upper(threshold) is deprecated, use clip(upper=threshold) instead
   int1  int2
0     1     1
1     1     1
```
**Before**
```python3
>>> data['float'] = data['int1'].astype(float)
>>> data.clip_upper(1)
   int1  int2  float
0   1.0   1.0    1.0
1   1.0   1.0    1.0
```
**After**
```python3
>>> data['float'] = data['int1'].astype(float)
>>> data.clip_upper(1)
   int1  int2  float
0     1     1    1.0
1     1     1    1.0
```

### CODE REVIEW

**Comment by jreback** [comment]
#20911 and #21476 maybe

**Comment by minggli** [comment]
> #20911 and #21476 maybe
Thanks @jreback sure let me check them out!

File: pandas/core/generic.py [hunk 1]
```diff
@@ -7132,28 +7132,21 @@ def _clip_with_scalar(self, lower, upper, inplace=False):
...
-            result[mask] = np.nan
+        result = self
 
-        axes_dict = self._construct_axes_dict()
-        result = self._constructor(result, **axes_dict).__finalize__(self)
+        if upper is not None:
+            subset = self.le(upper, axis=None) | isna(result)
+            result = result.where(subset, upper, axis=None, inplace=inplace)
+        if lower is not None:
+            if inplace:
```
Comment (mroeschke) [inline pandas/core/generic.py]: Why is this `inplace` check only when `lower is not None`? It looks like you also assign `result = self` before the upper and lower check.

**Comment by minggli** [comment]
@jreback,

had look at #20911, I think that's not a supported use case when passing a `DataFrame` for either `lower` or `upper`, which is essentially used as a numpy array for creating a mask, having ignored the column orders. 

#21476 is to do with different precisions between s and threshold. It works fine downcasting from int to float but appears not from float32 to float64. I suspect it's to do with numpy and need more investigation.

**Comment by minggli** [comment]
```python
>>> df.clip(lower=2, upper=7)
   floats  int
0     2.0    2
1     2.0    2
2     2.2    2
3     3.3    3
4     4.4    4
5     5.5    5
6     6.6    6
7     7.0    7
8     7.0    7
9     7.0    7

>>> df.clip(lower=2, upper=7.5)
   floats  int
0     2.0  2.0
1     2.0  2.0
2     2.2  2.0
3     3.3  3.0
4     4.4  4.0
5     5.5  5.0
6     6.6  6.0
7     7.5  7.0
8     7.5  7.5
9     7.5  7.5
```

File: pandas/core/generic.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/generic.py]: Got it. Thanks!

Comment (jreback) [reply pandas/core/generic.py]: I think its just better to use the existing idiom

```
if inplace:
     self._update_inplace(result0
else:
    return self
```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 1 above)
Comment (minggli) [reply pandas/core/generic.py]: actioned. thanks.

**Comment by minggli** [comment]
@jreback 

review actioned and added test and documentation.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks. ping on green.

**Comment by minggli** [comment]
@jreback  thanks, it's green.

**Comment by jreback** [comment]
thanks @minggli

**Comment by minggli** [comment]
@jreback @mroeschke thanks for reviewing and happy new year!

**Pull request merged**